_posts/junjun_zhang/2017-09-27-dag-db-with-version-control-in-elasticsearch.md

```diff
@@ -0,0 +1,340 @@
+
+
+## Implement it in Elasticsearch
+
+Now we are ready to put everything into Elasticsearch, you should be able to simply copy and paste
+the curl commands to run on your computer if you have Elasticsearch installed and accessible at
+`http://localhost:9200`.
+
+### Create version 1 of the DAG
+
+We will have two different kinds of Elasticsearch indexes, one for
+data, the other to keep list of graph member nodes' ID for each revision.
+
+First create JSON docs for each node in `DAG.1`. Here is an example to create `a1.1`:
+{% highlight bash %}
+# create a1.1, note that we do not need to have _id (a1.1) in the JSON document, the _id is specified in the HTTP request URL as document ID
+# we use '.d.' as namespace prefix to separate out from other indexes unrelated to our work
+curl -XPUT 'localhost:9200/.d.dag.data.a/A/a1.1' -d'
+{
+  "type": "A",
+  "id": "a1",
+  "p_id": [ ],
+  "a_id": [ ],
+  "data": { }
+}
+'
+{% endhighlight %}
+
+The full list of the curl commands to create all nodes is [here](https://gist.githubusercontent.com/junjun-zhang/8362c66d20f48073c644bd02062922c0/raw/c87fc94d2aac87e87e1bc2f0ceda0f4355b73e9f/1_create_nodes_for_dag_1.sh).
+
+Now, let's create JSON docs for graph member lists. Here is an example for `A.1`:
+{% highlight bash %}
+curl -XPUT 'localhost:9200/.d.dag.rev.a/A/A.1' -d'
+{ "m_ids": ["a1.1"] }
+'
+{% endhighlight %}
+
+Again, the full list of commands is [here](https://gist.githubusercontent.com/junjun-zhang/8362c66d20f48073c644bd02062922c0/raw/c87fc94d2aac87e87e1bc2f0ceda0f4355b73e9f/2_create_graph_member_lists_for_dag_1.sh).
+
+Finally, we need to link `DAG.1` to the collection of versioned node types: `["A.1", "B.1", "C.1", "D.1"]`.
+An elegant way to do this in Elasticsearch is to create an index alias with terms lookup
+filter to provide a virtual view to the underlying JSON docs with desired filtering. This
+is the core of the whole design.
+
+{% highlight bash %}
+curl -XPOST 'localhost:9200/_aliases' -d '
+{
+  "actions": [
+    {
+      "add": {
+        "index": ".d.dag.data.*",
+        "alias": ".d.dag.1",
+        "filter": {
+          "bool": { "minimum_should_match": 1,
+            "should": [ 
+              { "terms": { "_id": { "index": ".d.dag.rev.a", "type": "A", "id": "A.1", "path": "m_ids" } } },
+              { "terms": { "_id": { "index": ".d.dag.rev.b", "type": "B", "id": "B.1", "path": "m_ids" } } },
+              { "terms": { "_id": { "index": ".d.dag.rev.c", "type": "C", "id": "C.1", "path": "m_ids" } } },
+              { "terms": { "_id": { "index": ".d.dag.rev.d", "type": "D", "id": "D.1", "path": "m_ids" } } }
+            ]
+          }
+        }
+      }
+    }
+  ]
+}
+'
+{% endhighlight %}
+
+Version 1 of the DAG can always be access via the alias: `.d.dag.1`.
+
+### Create version 2 of the DAG
+
+Let's make changes to the DAG to make it version 2.
+
+Update `b1` from `b1.1` to `b1.2`, just need to create `b1.2`. We leave `b1.1` unchanged.
```

Comment:
Would it be possible here to have the example for a1 ? This way a reader can scroll up/down to see creation of a1 and update of a1 ? (instead of creating a1 and updating b1)

Reply:
That's a good suggestion. To allow visual comparison on the same node in different versions, I added the creation command for `b1.1` in the version 1 step. This is less involved than updating the examples to include change to a1.1.